embassy-stm32/src/dma/dma.rs

```diff
@@ -183,13 +188,43 @@ pub(crate) unsafe fn on_irq_inner(dma: pac::dma::Dma, channel_num: usize, index:
         panic!("DMA: error on DMA@{:08x} channel {}", dma.0 as u32, channel_num);
     }
 
-    if isr.tcif(channel_num % 4) && cr.read().tcie() {
-        /* acknowledge transfer complete interrupt */
-        dma.ifcr(channel_num / 4).write(|w| w.set_tcif(channel_num % 4, true));
+    let mut wake = false;
+
+    if isr.htif(channel_num % 4) && cr.read().htie() {
+        // Acknowledge half transfer complete interrupt
+        dma.ifcr(channel_num / 4).write(|w| w.set_htif(channel_num % 4, true));
+        wake = true;
+    }
+
+    wake |= process_tcif(dma, channel_num, index);
+
+    if wake {
         STATE.ch_wakers[index].wake();
     }
 }
 
+unsafe fn process_tcif(dma: pac::dma::Dma, channel_num: usize, index: usize) -> bool {
+    let isr_reg = dma.isr(channel_num / 4);
+    let cr_reg = dma.st(channel_num).cr();
+
+    // First, figure out if tcif is set without a cs.
+    if isr_reg.read().tcif(channel_num % 4) && cr_reg.read().tcie() {
+        // Make tcif test again within a cs to avoid race when incrementing complete_count.
```

Comment:
what race is this avoiding?

Reply:
The race is not possible if it is guaranteed that the DMA interrupt always have the highest priority.